### ISSUE TITLE
When synchronizing via snapshot, the problem with checking the first block if AppVersion is not equal to 0

### ISSUE THREAD

**Issue by klim0v** [tendermint/tendermint#7748]
The essence of the error is that the correct data restored from the snapshot gives an error when validating the block obtained from the code below, provided that AppVersion is not equal to zero:

[blockchain/v0/reactor.go](https://github.com/tendermint/tendermint/blob/v0.34.15/blockchain/v0/reactor.go#L353)
```
// See if there are any blocks to sync.
first, second := bcR.pool.PeekTwoBlocks()
  		
...

var err error
state, _, err = bcR.blockExec.ApplyBlock(state, firstID, first)
if err != nil {
  // TODO This is bad, are we zombie?
  panic(fmt.Sprintf("Failed to process committed block (%d:%X): %v", first.Height, first.Hash(), err))
}
```

The place where the error is generated:

[state/validation.go](https://github.com/tendermint/tendermint/blob/v0.34.15/state/validation.go#L22)
```
// Validate basic info.
if block.Version.App != state.Version.Consensus.App ||
	block.Version.Block != state.Version.Consensus.Block {
	return fmt.Errorf("wrong Block.Header.Version. Expected %v, got %v",
		state.Version.Consensus,
		block.Version,
	)
}
```

The structure in which the [block.Version.App](https://github.com/tendermint/tendermint/blob/41c176ccc6a75d25631d0f891efb2e19a33329dc/p2p/node_info.go#L60) field must be initialized at the time of the check is not filled in, because this is provided at the time of receiving the handshake, which is skipped. And validation is called after switch is executed [SwitchToFastSync](https://github.com/tendermint/tendermint/blob/41c176ccc6a75d25631d0f891efb2e19a33329dc/blockchain/v0/reactor.go#L125)

[node/node.go](https://github.com/tendermint/tendermint/blob/41c176ccc6a75d25631d0f891efb2e19a33329dc/node/node.go#L732)
```
   // Create the handshaker, which calls RequestInfo, sets the AppVersion on the state,
   // and replays any blocks as necessary to sync tendermint with the app.
   consensusLogger := logger.With("module", "consensus")
   if !stateSync {
   	if err := doHandshake(stateStore, state, blockStore, genDoc, eventBus, proxyApp, consensusLogger); err != nil {
   		return nil, err
   	}

   	// Reload the state. It will have the Version.Consensus.App set by the
   	// Handshake, and may have other modifications as well (ie. depending on
   	// what happened during block replay).
   	state, err = stateStore.Load()
   	if err != nil {
   		return nil, fmt.Errorf("cannot load state: %w", err)
   	}
   }
```

**Comment by klim0v** [comment]
As a temporary solution, I created a fork and commented out the error checking and throwing. This did not lead to new errors.

**Comment by lightyear15** [comment]
Hi @klim0v 
I came across same issue running an osmosis node with tendermint v.0.34.14 and was navigating through your analysis.
Tracing the error in current master branch I follow your trace up to ``internal/state/validation.go`` [``validateBlock``](https://github.com/tendermint/tendermint/blob/73f605af3fc778377e8f564086c72d261c669f88/internal/state/validation.go#L21)
To my view though, it looks the problem must be on the ``state`` side of the comparison, not on ``block`` side.
So I went that path and I tried to track down where the state comes from.
The state used in [``ApplyBlock``](https://github.com/tendermint/tendermint/blob/73f605af3fc778377e8f564086c72d261c669f88/internal/blocksync/reactor.go#L584) comes from ``r.InitialState`` which is initialized in [``SwitchToBlockSync``](https://github.com/tendermint/tendermint/blob/73f605af3fc778377e8f564086c72d261c669f88/internal/blocksync/reactor.go#L370), used in [``OnStart``](https://github.com/tendermint/tendermint/blob/73f605af3fc778377e8f564086c72d261c669f88/node/node.go#L415).
Here is a list of function calls splitted into callers and callees that traces the origin of the state
[caller](https://github.com/tendermint/tendermint/blob/73f605af3fc778377e8f564086c72d261c669f88/node/node.go#L506) - [callee](https://github.com/tendermint/tendermint/blob/73f605af3fc778377e8f564086c72d261c669f88/internal/statesync/reactor.go#L268)
[caller](https://github.com/tendermint/tendermint/blob/73f605af3fc778377e8f564086c72d261c669f88/internal/statesync/reactor.go#L314) - [callee](https://github.com/tendermint/tendermint/blob/73f605af3fc778377e8f564086c72d261c669f88/internal/statesync/syncer.go#L159)
[caller](https://github.com/tendermint/tendermint/blob/73f605af3fc778377e8f564086c72d261c669f88/internal/statesync/syncer.go#L208) - [callee](https://github.com/tendermint/tendermint/blob/73f605af3fc778377e8f564086c72d261c669f88/internal/statesync/syncer.go#L263)
The state if finally generated in [here](https://github.com/tendermint/tendermint/blob/73f605af3fc778377e8f564086c72d261c669f88/internal/statesync/syncer.go#L316)
I was about to declare "bug spotted" when I noticed [this line](https://github.com/tendermint/tendermint/blob/73f605af3fc778377e8f564086c72d261c669f88/internal/statesync/syncer.go#L356)
The bug might have been already fixed.

I don't know how to test it though :thinking:

**Comment by wwestgarth** [comment]
I'm seeing a similar issue when using state-sync with a non-zero `AppVersion` set. After a successful snapshot restore  tendermint crashes due to a divergence in `Block.Header.Version`:
```
I[2022-02-22|12:02:37.601] Applied snapshot chunk to ABCI app           module=statesync height=571 format=0 chunk=0 total=1
I[2022-02-22|12:02:37.878] Verified ABCI app                            module=statesync height=571 appHash=0D8699691F6337B104744FDCA6295CF13F5BCAFAE76AF79C60C411F290016341
I[2022-02-22|12:02:37.879] Snapshot restored                            module=statesync height=571 format=0 hash=0D8699691F6337B104744FDCA6295CF13F5BCAFAE76AF79C60C411F290016341
I[2022-02-22|12:02:37.884] Starting BlockPool service                   module=blockchain impl=BlockPool
panic: Failed to process committed block (572:FC77DE47BA9E46898205AAE4F5491FCA42956BEC8A895F09F4D7E784D2164C0F): wrong Block.Header.Version. Expected {11 2}, got {11 0}

goroutine 4825315 [running]:
github.com/tendermint/tendermint/blockchain/v0.(*BlockchainReactor).poolRoutine(0xc0019f81c0, 0x1)
	/Users/wwestgarth/go2/pkg/mod/github.com/tendermint/tendermint@v0.34.15/blockchain/v0/reactor.go:401 +0x1b09
created by github.com/tendermint/tendermint/blockchain/v0.(*BlockchainReactor).SwitchToFastSync
	/Users/wwestgarth/go2/pkg/mod/github.com/tendermint/tendermint@v0.34.15/blockchain/v0/reactor.go:125 +0x165
```

The interesting thing here (and what makes it differ from the linked cosmos SDK issue) is that the "expected" value has the correct value of `2` and the "got" value is zero. As this is from the viewpoint of the restored node, it means tht state-sync has the correct value, and is doing the correct thing and its the _other nodes_ pumping the wrong version into the block header.

In our application we are populating `ResponseInfo` with the following:
```
2022-02-22T11:48:50.611Z	DEBUG	processor	processor/abci.go:371	ABCI service INFO requested	{"height": 0, "hash": "", "appversion": 2, "software-version": "0.0.0"}
```

but in the tendermint logs we see that neither the app-version or the software-version is being picked up when the chain starts:
```
I[2022-02-22|11:48:46.622] ABCI Handshake App Info                      module=consensus height=0 hash= software-version= protocol-version=0
```

And if in my application I log the versions in the block header at the start of every block I see that they are zero:
```
2022-02-22T13:03:24.743Z	DEBUG	processor	processor/abci.go:545	entering begin block	{"at": "2022-02-22T13:03:24.743Z"}
2022-02-22T13:03:24.744Z	DEBUG	processor	processor/abci.go:546	Version	{"block": 11, "app": 0}
```

So for me, at least, state-sync is not at fault here -- the root cause is that in normal operation the `AppVersion` is never propagated into the block header.

**Comment by cmwaters** [comment]
This PR (https://github.com/tendermint/tendermint/pull/7885) should fix the app version problem for the v0.34 line. The stateprovider wasn't correctly setting the version when `State` was called. Oddly I thought I had done this as `master` and `v0.35` already have the fix but it seems it never reached v0.34. We should release v0.34.16 in the next week.
